File: go/restart_test.go
package signet

import (
	"context"
	"errors"
	"sync"
	"testing"
	"time"

	signetv1 "github.com/bytepunx/signet-clients/go/gen/signet/v1"
	"google.golang.org/protobuf/types/known/timestamppb"
)

// fakeLockStream is a hand-written fake satisfying lockStream, driven by a
// scripted sequence of responses (or errors) fed to Recv, and recording
// every Send for assertions.
type fakeLockStream struct {
	mu        sync.Mutex
	responses chan lockRecvResult
	sent      []*signetv1.AcquireRestartLockRequest
	closed    bool
}

type lockRecvResult struct {
	resp *signetv1.AcquireRestartLockResponse
	err  error
}

func newFakeLockStream(buf int) *fakeLockStream {
	return &fakeLockStream{responses: make(chan lockRecvResult, buf)}
}

func (f *fakeLockStream) pushResp(resp *signetv1.AcquireRestartLockResponse) {
	f.responses <- lockRecvResult{resp: resp}
}

func (f *fakeLockStream) pushErr(err error) {
	f.responses <- lockRecvResult{err: err}
}

func (f *fakeLockStream) Send(req *signetv1.AcquireRestartLockRequest) error {
	f.mu.Lock()
	defer f.mu.Unlock()
	if f.closed {
		return errors.New("send on closed stream")
	}
	f.sent = append(f.sent, req)
	return nil
}

func (f *fakeLockStream) Recv() (*signetv1.AcquireRestartLockResponse, error) {
	r, ok := <-f.responses
	if !ok {
		return nil, errors.New("stream closed")
	}
	return r.resp, r.err
}

func (f *fakeLockStream) CloseSend() error {
	f.mu.Lock()
	defer f.mu.Unlock()
	if f.closed {
		return nil
	}
	f.closed = true
	close(f.responses)
	return nil
}

func (f *fakeLockStream) sentCount() int {
	f.mu.Lock()
	defer f.mu.Unlock()
	return len(f.sent)
}

func acquiredResp(token string, expiresAt time.Time) *signetv1.AcquireRestartLockResponse {
	return &signetv1.AcquireRestartLockResponse{
		MessageType: signetv1.AcquireRestartLockResponse_MESSAGE_TYPE_ACQUIRED,
		Token:       token,
		ExpiresAt:   timestamppb.New(expiresAt),
	}
}

func queuePositionResp(pos int32) *signetv1.AcquireRestartLockResponse {
	return &signetv1.AcquireRestartLockResponse{
		MessageType: signetv1.AcquireRestartLockResponse_MESSAGE_TYPE_QUEUE_POSITION,
		Position:    pos,
	}
}

func ttlExtendedResp(expiresAt time.Time) *signetv1.AcquireRestartLockResponse {
	return &signetv1.AcquireRestartLockResponse{
		MessageType: signetv1.AcquireRestartLockResponse_MESSAGE_TYPE_TTL_EXTENDED,
		ExpiresAt:   timestamppb.New(expiresAt),
	}
}

func TestAcquireLock_RejectsNonPositiveTTL(t *testing.T) {
	if _, err := AcquireLock(context.Background(), nil, "ns", "svc", 0); err == nil {
		t.Fatal("expected error for ttl <= 0")
	}
	if _, err := AcquireLock(context.Background(), nil, "ns", "svc", -1); err == nil {
		t.Fatal("expected error for negative ttl")
	}
}

func TestAcquireLock_QueuePositionThenAcquired(t *testing.T) {
	stream := newFakeLockStream(4)
	stream.pushResp(queuePositionResp(2))
	stream.pushResp(queuePositionResp(1))
	stream.pushResp(acquiredResp("tok-1", time.Now().Add(30*time.Second)))

	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()

	lock, err := acquireLock(ctx, stream, "ns", "svc", 4*time.Second)
	if err != nil {
		t.Fatalf("acquireLock: %v", err)
	}
	defer lock.Release()

	if lock.Token() != "tok-1" {
		t.Errorf("Token() = %q, want tok-1", lock.Token())
	}
	if stream.sentCount() != 1 {
		t.Errorf("expected exactly 1 initial send before heartbeats, got %d", stream.sentCount())
	}
}

func TestAcquireLock_StreamErrorBeforeAcquired(t *testing.T) {
	stream := newFakeLockStream(4)
	wantErr := errors.New("boom")
	stream.pushErr(wantErr)

	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()

	_, err := acquireLock(ctx, stream, "ns", "svc", 4*time.Second)
	if err == nil {
		t.Fatal("expected error")
	}
}

func TestAcquireLock_HeartbeatIntervalIsTTLOverFour(t *testing.T) {
	stream := newFakeLockStream(4)
	stream.pushResp(acquiredResp("tok-1", time.Now().Add(4*time.Second)))

	ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
	defer cancel()

	// ttl=4s => heartbeat interval = 1s. Wait ~2.5s and expect >=2 heartbeats
	// sent (in addition to the 1 initial request).
	lock, err := acquireLock(ctx, stream, "ns", "svc", 4*time.Second)
	if err != nil {
		t.Fatalf("acquireLock: %v", err)
	}
	defer lock.Release()

	time.Sleep(2500 * time.Millisecond)
	if got := stream.sentCount(); got < 3 { // 1 initial + >=2 heartbeats
		t.Errorf("sentCount() = %d, want >= 3 (1 initial + >=2 heartbeats at ~1s interval)", got)
	}
}

func TestLock_TTLExtendedUpdatesExpiresAt(t *testing.T) {
	stream := newFakeLockStream(4)
	first := time.Now().Add(4 * time.Second)
	stream.pushResp(acquiredResp("tok-1", first))

	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()

	lock, err := acquireLock(ctx, stream, "ns", "svc", 4*time.Second)
	if err != nil {
		t.Fatalf("acquireLock: %v", err)
	}
	defer lock.Release()

	extended := time.Now().Add(60 * time.Second)
	stream.pushResp(ttlExtendedResp(extended))

	deadline := time.After(2 * time.Second)
	for {
		if lock.ExpiresAt().Equal(extended) {
			break
		}
		select {
		case <-deadline:
			t.Fatalf("ExpiresAt() never reflected TTL_EXTENDED; last = %v, want %v", lock.ExpiresAt(), extended)
		case <-time.After(10 * time.Millisecond):
		}
	}
}

func TestLock_ReleaseIsIdempotent(t *testing.T) {
	stream := newFakeLockStream(4)
	stream.pushResp(acquiredResp("tok-1", time.Now().Add(30*time.Second)))

	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()

	lock, err := acquireLock(ctx, stream, "ns", "svc", 4*time.Second)
	if err != nil {
		t.Fatalf("acquireLock: %v", err)
	}

	if err := lock.Release(); err != nil {
		t.Fatalf("first Release: %v", err)
	}
	if err := lock.Release(); err != nil {
		t.Fatalf("second Release should be a no-op, got: %v", err)
	}
}

func TestLock_LostReportedOnUnexpectedStreamError(t *testing.T) {
	stream := newFakeLockStream(4)
	stream.pushResp(acquiredResp("tok-1", time.Now().Add(4*time.Second)))

	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()

	lock, err := acquireLock(ctx, stream, "ns", "svc", 4*time.Second)
	if err != nil {
		t.Fatalf("acquireLock: %v", err)
	}
	defer lock.Release()

	wantErr := errors.New("lock lost")
	stream.pushErr(wantErr)

	select {
	case err := <-lock.Lost():
		if err == nil {
			t.Fatal("expected non-nil error on Lost()")
		}
	case <-time.After(2 * time.Second):
		t.Fatal("Lost() never signaled after stream error")
	}
}

func TestLock_ReleaseDoesNotReportLost(t *testing.T) {
	stream := newFakeLockStream(4)
	stream.pushResp(acquiredResp("tok-1", time.Now().Add(30*time.Second)))

	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()

	lock, err := acquireLock(ctx, stream, "ns", "svc", 4*time.Second)
	if err != nil {
		t.Fatalf("acquireLock: %v", err)
	}

	if err := lock.Release(); err != nil {
		t.Fatalf("Release: %v", err)
	}

	select {
	case err := <-lock.Lost():
		t.Fatalf("Lost() fired after a clean Release: %v", err)
	case <-time.After(200 * time.Millisecond):
		// expected: no loss signal for an intentional release
	}
}

// fakeWatchStream is a hand-written fake satisfying watchStream.
type fakeWatchStream struct {
	responses chan lockRecvResult2

	mu        sync.Mutex
	recvCount int
}

type lockRecvResult2 struct {
	resp *signetv1.WatchServiceBundleResponse
	err  error
}

func newFakeWatchStream(buf int) *fakeWatchStream {
	return &fakeWatchStream{responses: make(chan lockRecvResult2, buf)}
}

func (f *fakeWatchStream) pushChanged() {
	f.responses <- lockRecvResult2{resp: &signetv1.WatchServiceBundleResponse{
		EventType: signetv1.WatchServiceBundleResponse_EVENT_TYPE_CHANGED,
	}}
}

func (f *fakeWatchStream) pushErr(err error) {
	f.responses <- lockRecvResult2{err: err}
}

func (f *fakeWatchStream) recvCalls() int {
	f.mu.Lock()
	defer f.mu.Unlock()
	return f.recvCount
}

func (f *fakeWatchStream) Recv() (*signetv1.WatchServiceBundleResponse, error) {
	f.mu.Lock()
	f.recvCount++
	f.mu.Unlock()

	r, ok := <-f.responses
	if !ok {
		return nil, errors.New("stream closed")
	}
	return r.resp, r.err
}

func TestWatchLoop_CoalescesRapidChanges(t *testing.T) {
	// 3 changes are pushed but the response channel is never closed, so
	// after they're drained the 4th Recv() call just blocks — no error, no
	// reconnect, so open() is only ever called once.
	stream := newFakeWatchStream(8)
	stream.pushChanged()
	stream.pushChanged()
	stream.pushChanged()

	changes := make(chan struct{}, 1)
	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()

	go watchLoop(ctx, changes, func() (watchStream, error) {
		return stream, nil
	})

	// Wait until watchLoop has actually drained all 3 pushed events (the
	// 4th Recv call is the one that blocks) before checking coalescing —
	// otherwise this is racy: consuming the first signal before all 3 are
	// processed just frees the buffer for a legitimately new one.
	deadline := time.After(2 * time.Second)
	for stream.recvCalls() < 4 {
		select {
		case <-deadline:
			t.Fatalf("watchLoop only issued %d Recv calls, want >= 4", stream.recvCalls())
		case <-time.After(5 * time.Millisecond):
		}
	}

	if n := len(changes); n != 1 {
		t.Fatalf("len(changes) = %d after 3 rapid changes, want exactly 1 (coalesced)", n)
	}

	<-changes
	select {
	case <-changes:
		t.Fatal("expected no further pending signal after draining the coalesced one")
	default:
	}
}

func TestWatchLoop_ReconnectsAfterStreamError(t *testing.T) {
	first := newFakeWatchStream(2)
	first.pushErr(errors.New("boom"))

	second := newFakeWatchStream(2)
	second.pushChanged()

	var mu sync.Mutex
	attempt := 0

	changes := make(chan struct{}, 1)
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()

	go watchLoop(ctx, changes, func() (watchStream, error) {
		mu.Lock()
		defer mu.Unlock()
		attempt++
		if attempt == 1 {
			return first, nil
		}
		return second, nil
	})

	select {
	case <-changes:
	case <-time.After(4 * time.Second):
		t.Fatal("expected watchLoop to reconnect and eventually deliver a change")
	}
}

File: typescript/src/restart.test.ts
// Ports every case in ../../go/restart_test.go to the TypeScript client,
// plus a handful of cases specific to this port's design (AbortSignal
// cancellation, the dual Promise/EventEmitter 'lost' signal, backoff
// overrides for fast deterministic reconnect tests). All cases run against
// hand-written fakes implementing LockStream/WatchStream — no real network
// connection or signet instance is involved anywhere in this file.
import assert from "node:assert/strict";
import { test } from "node:test";
import {
  AcquireRestartLockResponse_MessageType,
  type AcquireRestartLockRequest,
  type AcquireRestartLockResponse,
  WatchServiceBundleResponse_EventType,
  type WatchServiceBundleResponse,
} from "./gen/signet/v1/secrets.js";
import { BundleWatch, Lock, acquireLockWithStream, runWatchLoop, type LockStream, type WatchStream } from "./restart.js";
import { acquireLock } from "./restart.js";

// ---------------------------------------------------------------------------
// Fakes (mirrors go/restart_test.go's fakeLockStream/fakeWatchStream)
// ---------------------------------------------------------------------------

type QueueItem<T> = { resp: T } | { err: Error };

class FakeLockStream implements LockStream {
  readonly sent: AcquireRestartLockRequest[] = [];
  private readonly queue: QueueItem<AcquireRestartLockResponse>[] = [];
  private readonly waiters: Array<(item: QueueItem<AcquireRestartLockResponse>) => void> = [];
  private closed = false;

  pushResp(resp: AcquireRestartLockResponse): void {
    this.push({ resp });
  }

  pushErr(err: Error): void {
    this.push({ err });
  }

  private push(item: QueueItem<AcquireRestartLockResponse>): void {
    const waiter = this.waiters.shift();
    if (waiter) waiter(item);
    else this.queue.push(item);
  }

  send(req: AcquireRestartLockRequest): void {
    if (this.closed) throw new Error("send on closed stream");
    this.sent.push(req);
  }

  recv(): Promise<AcquireRestartLockResponse> {
    return new Promise((resolve, reject) => {
      const deliver = (item: QueueItem<AcquireRestartLockResponse>) => {
        if ("err" in item) reject(item.err);
        else resolve(item.resp);
      };
      if (this.queue.length > 0) {
        deliver(this.queue.shift()!);
      } else if (this.closed) {
        reject(new Error("stream closed"));
      } else {
        this.waiters.push(deliver);
      }
    });
  }

  closeSend(): void {
    if (this.closed) return;
    this.closed = true;
    while (this.waiters.length > 0) {
      this.waiters.shift()!({ err: new Error("stream closed") });
    }
  }

  sentCount(): number {
    return this.sent.length;
  }
}

class FakeWatchStream implements WatchStream {
  recvCalls = 0;
  private readonly queue: QueueItem<WatchServiceBundleResponse>[] = [];
  private readonly waiters: Array<(item: QueueItem<WatchServiceBundleResponse>) => void> = [];

  pushChanged(): void {
    this.push({
      resp: {
        eventType: WatchServiceBundleResponse_EventType.EVENT_TYPE_CHANGED,
        namespace: "",
        service: "",
      },
    });
  }

  pushErr(err: Error): void {
    this.push({ err });
  }

  private push(item: QueueItem<WatchServiceBundleResponse>): void {
    const waiter = this.waiters.shift();
    if (waiter) waiter(item);
    else this.queue.push(item);
  }

  recv(): Promise<WatchServiceBundleResponse> {
    this.recvCalls++;
    return new Promise((resolve, reject) => {
      const deliver = (item: QueueItem<WatchServiceBundleResponse>) => {
        if ("err" in item) reject(item.err);
        else resolve(item.resp);
      };
      if (this.queue.length > 0) deliver(this.queue.shift()!);
      else this.waiters.push(deliver);
    });
  }
}

function acquiredResp(token: string, expiresAt: Date): AcquireRestartLockResponse {
  return {
    messageType: AcquireRestartLockResponse_MessageType.MESSAGE_TYPE_ACQUIRED,
    position: 0,
    token,
    expiresAt,
  };
}

function queuePositionResp(position: number): AcquireRestartLockResponse {
  return {
    messageType: AcquireRestartLockResponse_MessageType.MESSAGE_TYPE_QUEUE_POSITION,
    position,
    token: "",
    expiresAt: undefined,
  };
}

function ttlExtendedResp(expiresAt: Date): AcquireRestartLockResponse {
  return {
    messageType: AcquireRestartLockResponse_MessageType.MESSAGE_TYPE_TTL_EXTENDED,
    position: 0,
    token: "",
    expiresAt,
  };
}

/** Resolves to the promise's value, or to `sentinel` if it doesn't settle within ms. */
function raceTimeout<T>(promise: Promise<T>, ms: number, sentinel: symbol): Promise<T | symbol> {
  return Promise.race([promise, new Promise<symbol>((resolve) => setTimeout(() => resolve(sentinel), ms))]);
}

const TIMEOUT = Symbol("timeout");

async function waitUntil(predicate: () => boolean, timeoutMs: number, message: string): Promise<void> {
  const deadline = Date.now() + timeoutMs;
  while (!predicate()) {
    if (Date.now() > deadline) throw new Error(message);
    await new Promise((resolve) => setTimeout(resolve, 5));
  }
}

// ---------------------------------------------------------------------------
// acquireLock / acquireLockWithStream
// ---------------------------------------------------------------------------

test("acquireLock rejects ttl <= 0 before opening any stream", async () => {
  await assert.rejects(() => acquireLock(undefined as never, "ns", "svc", 0), /ttlSeconds must be > 0, got 0/);
  await assert.rejects(() => acquireLock(undefined as never, "ns", "svc", -1), /ttlSeconds must be > 0, got -1/);
});

test("acquireLock queue position then acquired handoff", async () => {
  const stream = new FakeLockStream();
  stream.pushResp(queuePositionResp(2));
  stream.pushResp(queuePositionResp(1));
  stream.pushResp(acquiredResp("tok-1", new Date(Date.now() + 30_000)));

  const lock = await acquireLockWithStream(stream, "ns", "svc", 4);
  try {
    assert.equal(lock.token, "tok-1");
    assert.equal(stream.sentCount(), 1, "expected exactly 1 initial send before heartbeats");
  } finally {
    await lock.release();
  }
});

test("acquireLock surfaces a clear error when the stream errors before ACQUIRED", async () => {
  const stream = new FakeLockStream();
  stream.pushErr(new Error("boom"));

  await assert.rejects(
    () => acquireLockWithStream(stream, "ns", "svc", 4),
    /AcquireRestartLock stream ended before the lock was acquired.*boom/,
  );
});

test("heartbeat interval defaults to ttlSeconds / 4", async () => {
  const stream = new FakeLockStream();
  stream.pushResp(acquiredResp("tok-1", new Date(Date.now() + 2_000)));

  // ttl=2s => heartbeat interval = 500ms. Wait ~1.3s and expect >=2
  // heartbeats sent (in addition to the 1 initial request).
  const lock = await acquireLockWithStream(stream, "ns", "svc", 2);
  try {
    await new Promise((resolve) => setTimeout(resolve, 1300));
    assert.ok(stream.sentCount() >= 3, `sentCount() = ${stream.sentCount()}, want >= 3 (1 initial + >=2 heartbeats)`);
    const heartbeats = stream.sent.slice(1);
    for (const hb of heartbeats) {
      assert.equal(hb.heartbeat, true);
    }
  } finally {
    await lock.release();
  }
});

test("TTL_EXTENDED updates the tracked expiry", async () => {
  const stream = new FakeLockStream();
  const first = new Date(Date.now() + 4_000);
  stream.pushResp(acquiredResp("tok-1", first));

  const lock = await acquireLockWithStream(stream, "ns", "svc", 4);
  try {
    const extended = new Date(Date.now() + 60_000);
    stream.pushResp(ttlExtendedResp(extended));

    await waitUntil(
      () => lock.expiresAt?.getTime() === extended.getTime(),
      2000,
      `expiresAt never reflected TTL_EXTENDED; last = ${lock.expiresAt?.toISOString()}, want ${extended.toISOString()}`,
    );
  } finally {
    await lock.release();
  }
});

test("release is idempotent", async () => {
  const stream = new FakeLockStream();
  stream.pushResp(acquiredResp("tok-1", new Date(Date.now() + 30_000)));

  const lock = await acquireLockWithStream(stream, "ns", "svc", 4);
  await lock.release();
  await lock.release(); // must not throw / hang
});

test("lock loss is reported on an unexpected stream error", async () => {
  const stream = new FakeLockStream();
  stream.pushResp(acquiredResp("tok-1", new Date(Date.now() + 4_000)));

  const lock = await acquireLockWithStream(stream, "ns", "svc", 4);
  try {
    const eventErr: Promise<Error> = new Promise((resolve) => lock.once("lost", resolve));

    stream.pushErr(new Error("lock lost"));

    const err = await raceTimeout(lock.lost, 2000, TIMEOUT);
    assert.notEqual(err, TIMEOUT, "lost never signaled after stream error");
    assert.ok(err instanceof Error);
    assert.match((err as Error).message, /lock lost/);

    // The 'lost' EventEmitter event must fire with the same error too.
    const emitted = await raceTimeout(eventErr, 500, TIMEOUT);
    assert.notEqual(emitted, TIMEOUT, "'lost' event never emitted");
  } finally {
    await lock.release();
  }
});

test("release does not report lost", async () => {
  const stream = new FakeLockStream();
  stream.pushResp(acquiredResp("tok-1", new Date(Date.now() + 30_000)));

  const lock = await acquireLockWithStream(stream, "ns", "svc", 4);
  await lock.release();

  const result = await raceTimeout(lock.lost, 200, TIMEOUT);
  assert.equal(result, TIMEOUT, "lost fired after a clean release()");
});

test("acquireLock rejects when the signal aborts while waiting for ACQUIRED", async () => {
  const stream = new FakeLockStream();
  // Never push a response — the caller aborts while still waiting in queue.
  const controller = new AbortController();
  const promise = acquireLockWithStream(stream, "ns", "svc", 4, { signal: controller.signal });
  controller.abort(new Error("shutting down"));

  await assert.rejects(() => promise, /AcquireRestartLock.*aborted.*shutting down/);
});

// ---------------------------------------------------------------------------
// WatchBundle / runWatchLoop
// ---------------------------------------------------------------------------

test("watch loop coalesces rapid successive changes into one pending signal", async () => {
  // 3 changes are pushed but the stream never errors or ends, so after
  // they're drained the 4th recv() call just hangs — no error, no
  // reconnect, so open() is only ever called once.
  const stream = new FakeWatchStream();
  stream.pushChanged();
  stream.pushChanged();
  stream.pushChanged();

  const watch = new BundleWatch();
  void runWatchLoop(() => stream, watch, {});

  await waitUntil(
    () => stream.recvCalls >= 4,
    2000,
    `watch loop only issued ${stream.recvCalls} recv calls, want >= 4`,
  );

  const first = await raceTimeout(watch.next(), 500, TIMEOUT);
  assert.equal(first, true, "expected the coalesced signal to be immediately available");

  const second = await raceTimeout(watch.next(), 200, TIMEOUT);
  assert.equal(second, TIMEOUT, "expected no further pending signal after draining the coalesced one");
});

test("watch loop reconnects with backoff after a stream error and eventually delivers a change", async () => {
  const first = new FakeWatchStream();
  first.pushErr(new Error("boom"));

  const second = new FakeWatchStream();
  second.pushChanged();

  let attempt = 0;
  const open = () => {
    attempt++;
    return attempt === 1 ? first : second;
  };

  const watch = new BundleWatch();
  // Fast backoff so this test doesn't take 1s+ in CI; the *default*
  // 1s/30s backoff constants are exercised implicitly by every other
  // reconnect-shaped test using default options.
  void runWatchLoop(open, watch, { backoffMinMs: 10, backoffMaxMs: 50 });

  const result = await raceTimeout(watch.next(), 2000, TIMEOUT);
  assert.equal(result, true, "expected watch loop to reconnect and eventually deliver a change");
});

test("watch loop treats a failing open() the same as a stream error (backs off and retries)", async () => {
  let attempt = 0;
  const open = () => {
    attempt++;
    if (attempt === 1) throw new Error("dial refused");
    const stream = new FakeWatchStream();
    stream.pushChanged();
    return stream;
  };

  const watch = new BundleWatch();
  void runWatchLoop(open, watch, { backoffMinMs: 10, backoffMaxMs: 50 });

  const result = await raceTimeout(watch.next(), 2000, TIMEOUT);
  assert.equal(result, true, "expected watch loop to retry after a failed open() and eventually deliver a change");
  assert.ok(attempt >= 2);
});

test("watch loop stops and closes the BundleWatch when the signal is aborted", async () => {
  const stream = new FakeWatchStream(); // never delivers anything
  const controller = new AbortController();

  const watch = new BundleWatch();
  const loopDone = runWatchLoop(() => stream, watch, { signal: controller.signal });

  controller.abort();
  await loopDone;

  const result = await raceTimeout(watch.next(), 500, TIMEOUT);
  assert.equal(result, false, "expected next() to resolve false once the watch is closed");
});

// ---------------------------------------------------------------------------
// Lock.token / Lock.expiresAt basic accessors (not covered by the flows above)
// ---------------------------------------------------------------------------

test("Lock exposes token and expiresAt from the ACQUIRED response", async () => {
  const stream = new FakeLockStream();
  const expiresAt = new Date(Date.now() + 12_000);
  stream.pushResp(acquiredResp("tok-xyz", expiresAt));

  const lock: Lock = await acquireLockWithStream(stream, "ns", "svc", 8);
  try {
    assert.equal(lock.token, "tok-xyz");
    assert.equal(lock.expiresAt?.getTime(), expiresAt.getTime());
  } finally {
    await lock.release();
  }
});
